zone/doors.cpp

```diff
@@ -190,6 +190,204 @@ bool Doors::Process()
+			if (ourZoneInstanceID == GUILD_NONE && leader_instanceid != GUILD_NONE)
+			{
+				zoneguildid = leader_instanceid;
+
+				int64 curzoneLockout = zone->GetZoneInstanceLockoutByCharacterAndZone(sender->CharacterID(), zoneid);
+				//TODO: hardcode, retrieve from db later
+				int64 zonelockout = RuleI(Quarm, InstanceMinimumLockoutTime);
+
+				auto cur_time = time(nullptr);
+
+				int64 end_time = cur_time + RuleI(Quarm, InstanceMinimumLockoutTime);
+				if (curzoneLockout <= end_time)
+				{
+					zone->ReplaceZoneInstanceIDCache(sender->CharacterID(), zoneid, zoneguildid, end_time);
+					database.SaveCharacterInstanceLockout(sender->CharacterID(), end_time, zoneid, zoneguildid);
+					CharacterInstanceLockout instanceLockout;
+					memset(&instanceLockout, 0, sizeof(CharacterInstanceLockout));
+					instanceLockout.character_id = sender->CharacterID();
+					instanceLockout.expirydate = end_time;
+					instanceLockout.zone_id = zoneid;
+					instanceLockout.zone_instance_id = zoneguildid;
+					sender->character_instance_lockouts[zoneid] = instanceLockout;
+				}
+			}
+			else
+			{
+				sender->Message(Chat::Red, "You are unable to enter an instance because your raid leader hasn't entered first.");
+				return false;
+			}
+		}
+		else
+		{
+			uint32 leader_instanceid = zone->GetZoneInstanceIDByCharacterAndZone(player_raid->GetRaidLeaderCharacterID(), zoneid);
+			if (ourZoneInstanceID == GUILD_NONE)
+			{
+
+				zoneguildid = database.GetHighestZoneInstanceID(zoneid) + 1;
+				//TODO: hardcode, retrieve from db later
+				int64 zonelockout = RuleI(Quarm, InstanceMinimumLockoutTime);
+				auto cur_time = time(nullptr);
+				int64 end_time = cur_time + RuleI(Quarm, InstanceMinimumLockoutTime);
+				int64 curzoneLockout = zone->GetZoneInstanceLockoutByCharacterAndZone(sender->CharacterID(), zoneid);
+				if (curzoneLockout <= end_time)
+				{
+					zone->ReplaceZoneInstanceIDCache(sender->CharacterID(), zoneid, zoneguildid, end_time);
+					database.SaveCharacterInstanceLockout(sender->CharacterID(), end_time, zoneid, zoneguildid);
+					CharacterInstanceLockout instanceLockout;
+					memset(&instanceLockout, 0, sizeof(CharacterInstanceLockout));
+					instanceLockout.character_id = sender->CharacterID();
+					instanceLockout.expirydate = end_time;
+					instanceLockout.zone_id = zoneid;
+					instanceLockout.zone_instance_id = zoneguildid;
+					sender->character_instance_lockouts[zoneid] = instanceLockout;
+				}
+			}
+		}
+	}
+
+	if (zoneguildid == GUILD_NONE)
+	{
+		sender->Message(Chat::Red, "You are unable to enter this instance.");
+		return false;
+	}
+
+	if (strncmp(destination_zone_name, zone_name, strlen(zone_name)) == 0 && !keyneeded)
+	{
+		if (!keepoffkeyring)
+		{
+			sender->KeyRingAdd(key);
+		}
+		sender->MovePCGuildID(zone->GetZoneID(), zoneguildid, m_destination.x, m_destination.y, m_destination.z, m_destination.w);
```

Comment:
Sometimes zone->GetZoneID() is being used and sometimes zoneid. Are they going to be different? If not maybe good to make them consistent.

Reply:
This is copied from the below logic. It'd be a separate refactor. Zoning code is fickle and I am hesitant to touch what works